py-polars/polars/expr/list.py

```diff
@@ -1092,7 +1092,7 @@ def to_array(self, width: int) -> Expr:
 
     def to_struct(
         self,
-        n_field_strategy: ToStructStrategy = "first_non_null",
+        n_field_strategy: ListToStructWidthStrategy = "first_non_null",
         fields: Sequence[str] | Callable[[int], str] | None = None,
```

Comment:
We have this functionality where you can pass a Python function that gets called with a field index to generate names, but I'm not sure it's needed since you can just pass a list of field names. Maybe we should remove it?

Reply:
It allows you to create fields at runtime. Which is a bit dangerous as it will be dependent of the data. But I think we should keep for now.

I want to add a `return_dtype: DataType` argument as well that allows you to specify the datatype immediately.

Reply:
Can we add a `Notes` warning that it is a bug in a users query if they don't set `fields` or `return_dtype`? Then we can make it required in 2.0.

Reply:
> I want to add a `return_dtype: DataType` argument as well that allows you to specify the datatype immediately.

I think we can already do this by passing a list of names to the `fields` parameter. For the function case, we also get the output schema if the `upper_bound` is set - I can add a warning if we get a function without `upper_bound` being set?

Reply:
I have added these warnings -

```python
DeprecationWarning: `fields` should be specified when calling list.to_struct, otherwise the output schema will not be known, causing subsequent operations to fail. Future versions of Polars may be changed to raise an error in these cases.
  print(lf.select(pl.first().list.to_struct(upper_bound=1)).collect_schema())

DeprecationWarning: `upper_bound` should be specified when calling list.to_struct with a function as `fields`, otherwise the output schema will not be known, causing subsequent operations to fail. Future versions of Polars may be changed to raise an error in these cases.
  print(lf.select(pl.first().list.to_struct(fields=str)).collect_schema())
```

Reply:
Actually, I think I want to add the warnings/notes in a separate PR